src/main/java/org/broadinstitute/hellbender/utils/read/markduplicates/PairedEnds.java

```diff
@@ -1,130 +1,441 @@
-  public void setReadGroup(final short readGroup) { this.readGroup = readGroup; }
+            if( read1UnclippedStart < read2UnclippedStart ){
+                first = read1;
+                second = read2;
+                firstUnclippedStartPosition = read1UnclippedStart;
+                secondUnclippedStartPosition = read2UnclippedStart;
+            } else {
+                first = read2;
+                second = read1;
+                firstUnclippedStartPosition = read2UnclippedStart;
+                secondUnclippedStartPosition = read1UnclippedStart;
+            }
 
-  @Override
-  public short getTile() { return this.tile; }
+            firstStartPosition = first.getAssignedStart();
+            firstRefIndex = (short)ReadUtils.getReferenceIndex(first, header);
+            secondRefIndex = (short)ReadUtils.getReferenceIndex(second, header);
 
-  @Override
-  public void setTile(final short tile) { this.tile = tile; }
+            final GATKRead firstOfPair;
+            final GATKRead secondOfPair;
+            if (read1.isFirstOfPair()){
+                firstOfPair = read1;
+                secondOfPair = read2;
+            } else {
+                firstOfPair = read2;
+                secondOfPair = read1;
+            }
 
-  @Override
-  public short getX() { return this.x; }
+            R1R = firstOfPair.isReverseStrand();
+            R2R = secondOfPair.isReverseStrand();
+        }
 
-  @Override
-  public void setX(final short x) { this.x = x; }
+        // Constructor for serialization purposes
+        private Pair(Kryo kryo, Input input){
+            super(input.readInt(true), input.readString());
+            first = null;
+            second = null;
 
-  @Override
-  public short getY() { return this.y; }
+            // Information used to detect optical dupes
+            readGroup = -1;
+            tile = -1;
+            x = -1;
+            y = -1;
+            libraryId = -1;
 
-  @Override
-  public void setY(final short y) { this.y = y; }
+            score = input.readInt();
 
-  @Override
-  public short getLibraryId() { return this.libraryId; }
+            firstStartPosition = input.readInt();
+            firstUnclippedStartPosition = input.readInt();
+            firstRefIndex = input.readShort();
+            R1R = input.readBoolean();
 
-  @Override
-  public void setLibraryId(final short libraryId) { this.libraryId = libraryId; }
-  /**
-   * returns a deep(ish) copy of the GATK reads in the PairedEnds.
-   * TODO: This is only deep for the Google Model read, GATKRead copy() isn't deep for
-   * TODO: for the SAMRecord backed read.
-   * @return a new deep copy
-   */
-  public PairedEnds copy() {
-    if (second == null) {
-      return new PairedEnds(first.copy());
+            secondUnclippedStartPosition = input.readInt();
+            secondRefIndex = input.readShort();
+            R2R = input.readBoolean();
+
+        }
+
+        protected void serialize(Kryo kryo, Output output) {
```

Comment:
we can drop the kryo from this method since I don't think we ever use it

Reply:
done